src/java/org/apache/cassandra/net/BufferPoolAllocator.java

```diff
@@ -92,18 +104,47 @@ void release()
     public static class Wrapped extends UnpooledUnsafeDirectByteBuf
     {
         private ByteBuffer wrapped;
+        private boolean pooled = true;
 
-        Wrapped(BufferPoolAllocator allocator, ByteBuffer wrap)
+        Wrapped(BufferPoolAllocator allocator, ByteBuffer wrap, int maxCapacity)
         {
-            super(allocator, wrap, wrap.capacity());
+            super(allocator, wrap, max(wrap.capacity(), maxCapacity));
             wrapped = wrap;
         }
 
+        @Override
+        public ByteBuf capacity(int newCapacity)
+        {
+            if (newCapacity == capacity())
+                return this;
+
+            // resizing doesn't use the pool
+            ByteBuf newBuffer = super.capacity(newCapacity);
+            ByteBuffer nioBuffer = newBuffer.nioBuffer(0, newBuffer.capacity());
+
+            if (pooled)
+                bufferPool.put(wrapped);
+
+            wrapped = nioBuffer;
+            pooled = false;
+            return newBuffer;
+        }
+
         @Override
         public void deallocate()
         {
-            if (wrapped != null)
+            if (wrapped == null)
+                return;
+
+            // release initial buffer from the pool
+            if (pooled)
                 bufferPool.put(wrapped);
+            // or, if resized, release the new buffer
```

Comment:
@jasonstack From your last feedback in the old PR, I added this piece of code to release the resized buffer.